File: src/lib/scala/Slate.Test/src/test/scala/slate/test/MapperTests.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2016 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */
package slate.test

import scala.reflect.runtime.universe.{typeOf,Type}
import org.scalatest.{BeforeAndAfterAll, BeforeAndAfter, FunSpec}
import slate.common.Random._
import slate.common.{ListMap, DateTime, Field}
import slate.common.mapper.{MapperSourceReaderMap, Mapper}
import slate.entities.core.EntityMapper

import scala.annotation.meta.field

class MapperTests  extends FunSpec  with BeforeAndAfter with BeforeAndAfterAll {

  before {
    println("before")
  }


  describe ( "Case Classes") {

    it("can register by key") {
      val mod = new AuthorR(email = "kishore@abc.com")
      val model = Mapper.loadSchema(typeOf[AuthorR])

      model.fields.foreach(println)

      val mapper = new EntityMapper(model)
      val sqlCreate = mapper.mapToSql(mod, false, true)
      println(sqlCreate)

      val sqlUpdate = mapper.mapToSql(mod, true, true)
      println(sqlUpdate)
    }


    it("can map to case class") {
      val mod = new AuthorR(email = "kishore@abc.com")
      val model = Mapper.loadSchema(typeOf[AuthorR])

      val mapper = new EntityMapper(model)
      val data = new ListMap[String,Any]()
      data.add("id", 1L)
      data.add("createdAt", new java.sql.Timestamp(2017, 1, 1, 12,0,0,0))
      data.add("createdBy", 100L)
      data.add("updatedAt", new java.sql.Timestamp(2017, 1, 2, 12,0,0,0))
      data.add("updatedBy", 101L)
      data.add("uniqueId", "ABC")
      data.add("email", "kishore@abc.com")
      data.add("isActive", true)
      data.add("age", 35)

      val source = new MapperSourceReaderMap(data)
      val entity = mapper.mapFrom(source).get.asInstanceOf[AuthorR]

      assert( entity.id == 1L )
      assert( entity.uniqueId == "ABC" )
      assert( entity.email == "kishore@abc.com" )
      assert( entity.isActive == true )
      assert( entity.age == 35 )
    }


    it("can map to var class") {
      val mod = new AuthorW()
      mod.email = "kishore@abc.com"
      val model = Mapper.loadSchema(typeOf[AuthorW])

      val mapper = new EntityMapper(model)
      val data = new ListMap[String,Any]()
      data.add("id", 1L)
      data.add("createdAt", new java.sql.Timestamp(2017, 1, 1, 12,0,0,0))
      data.add("createdBy", 100L)
      data.add("updatedAt", new java.sql.Timestamp(2017, 1, 2, 12,0,0,0))
      data.add("updatedBy", 101L)
      data.add("uniqueId", "ABC")
      data.add("email", "kishore@abc.com")
      data.add("isActive", true)
      data.add("age", 35)

      val source = new MapperSourceReaderMap(data)
      val entity = mapper.mapFrom(source).get.asInstanceOf[AuthorW]

      assert( entity.id == 1L )
      assert( entity.uniqueId == "ABC" )
      assert( entity.email == "kishore@abc.com" )
      assert( entity.isActive == true )
      assert( entity.age == 35 )
    }
  }
}


case class AuthorR
(
      id       : Long             = 0,

      @(Field@field)("", true, -1)
      createdAt: DateTime         = DateTime.now(),

      @(Field@field)("", true, -1)
      createdBy: Long             = 0,

      @(Field@field)("", true, -1)
      updatedAt: DateTime         = DateTime.now(),

      @(Field@field)("", true, -1)
      updatedBy: Long             = 0,

      @(Field@field)("", true, -1)
      uniqueId: String            = stringGuid(),

      @(Field@field)("", true, 30)
      email:String = "",

      @(Field@field)("", true, 30)
      isActive:Boolean = false,

      @(Field@field)("", true, 0)
      age:Int = 35
)


class AuthorW {
      var id: Long = 0

      @Field("", true, -1)
      var createdAt: DateTime = DateTime.now()

      @Field("", true, -1)
      var createdBy: Long = 0

      @Field("", true, -1)
      var updatedAt: DateTime = DateTime.now()

      @Field("", true, -1)
      var updatedBy: Long = 0

      @Field("", true, -1)
      var uniqueId: String = stringGuid()

      @Field("", true, 30)
      var email: String = ""

      @Field("", true, 30)
      var isActive: Boolean = false

      @Field("", true, 0)
      var age: Int = 35

}

File: src/lib/kotlin/slatekit-tests/src/test/kotlin/test/MapperTests.kt
/**
<slate_header>
url: www.slatekit.com
git: www.github.com/code-helix/slatekit
org: www.codehelix.co
author: Kishore Reddy
copyright: 2016 CodeHelix Solutions Inc.
license: refer to website and/or github
about: A Kotlin utility library, tool-kit and server backend.
mantra: Simplicity above all else
</slate_header>
 */
package test

import org.junit.Test
import slatekit.common.Random
import slatekit.common.DateTime
import slatekit.common.Field
import slatekit.common.ListMap
import slatekit.common.Mapper
import slatekit.common.records.RecordMap
import slatekit.entities.core.EntityMapper
import slatekit.meta.models.Model
import slatekit.meta.models.ModelMapper
import kotlin.reflect.KClass

/**
 * Created by kishorereddy on 6/4/17.
 */

class MapperTests {

    fun loadSchemaViaReflection(): Model {
        val model = ModelMapper.loadSchema(AuthorR::class)
        return model
    }


    fun loadSchemaSpecification(): Model {
        val model = Model(AuthorR::class)
                    .addId( AuthorR::id, true)
                    .add( AuthorR::createdAt )
                    .add( AuthorR::createdBy )
                    .add( AuthorR::updatedAt )
                    .add( AuthorR::updatedBy )
                    .add( AuthorR::uniqueId  )
                    .add( AuthorR::email     )
                    .add( AuthorR::isActive  )
                    .add( AuthorR::age       )
                    .add( AuthorR::salary    )
        return model
    }


    fun ensureField(model:Model, name:String, required:Boolean, tpe: KClass<*>):Unit {
        val field = model.fields.find { it.name == name }
        assert(field != null)
        assert(field!!.isRequired == required)
        assert(field!!.dataType == tpe)
    }


    fun ensureAuthorModel(model:Model):Unit {

        assert(model.hasId)
        assert(model.any)
        ensureField(model, "id"        , true, Long::class     )
        ensureField(model, "createdAt" , true, DateTime::class )
        ensureField(model, "createdBy" , true, Long::class     )
        ensureField(model, "updatedAt" , true, DateTime::class )
        ensureField(model, "updatedBy" , true, Long::class     )
        ensureField(model, "uniqueId"  , true, String::class   )
        ensureField(model, "email"     , true, String::class   )
        ensureField(model, "isActive"  , true, Boolean::class  )
        ensureField(model, "age"       , true, Int::class      )
        ensureField(model, "salary"    , true, Double::class   )
    }


    @Test fun can_build_from_reflection(){
        val model = loadSchemaViaReflection()
        ensureAuthorModel(model)
    }


    @Test fun can_build_from_schema(){
        val model = loadSchemaSpecification()
        ensureAuthorModel(model)
    }


    @Test fun can_map_to_case_class(){
        val mod = AuthorR(email = "kishore@abc.com")
        val model = loadSchemaViaReflection()

        val mapper = EntityMapper(model)
        val data = ListMap(listOf(
        Pair("id", 1L),
        Pair("createdAt", java.sql.Timestamp(2017, 1, 1, 12,0,0,0)),
        Pair("createdBy", 100L),
        Pair("updatedAt", java.sql.Timestamp(2017, 1, 2, 12,0,0,0)),
        Pair("updatedBy", 101L),
        Pair("uniqueId", "ABC"),
        Pair("email", "kishore@abc.com"),
        Pair("isActive", true),
        Pair("age", 35),
        Pair("salary", 400.5)
        ))

        val source = RecordMap(data)
        val entity = mapper.mapFrom(source)!!as AuthorR

        assert( entity.id == 1L )
        assert( entity.uniqueId == "ABC" )
        assert( entity.email == "kishore@abc.com" )
        assert( entity.isActive )
        assert( entity.age == 35 )
        assert( entity.salary == 400.5 )
    }


    data class AuthorR(
            val id: Long             = 0,

            @property:Field(required = true)
            val createdAt: DateTime = DateTime.now(),

            @property:Field(required = true)
            val createdBy: Long = 0,

            @property:Field(required = true)
            val updatedAt: DateTime = DateTime.now(),

            @property:Field(required = true)
            val updatedBy: Long = 0,

            @property:Field(required = true)
            val uniqueId: String            = Random.stringGuid(),

            @property:Field(required = true)
            val email:String = "",

            @property:Field(required = true)
            val isActive:Boolean = false,

            @property:Field(required = true)
            val age:Int = 35,

            @property:Field(required = true)
            val salary:Double = 20.5
    )
}